### PR TITLE
 Add kubernetes secret resource (operation handlers), data source and high level schema with validations

### PR DESCRIPTION
**Pull request by Shubbhang351**
1. **What this PR does / why we need it**:
     Add kubernetes secret resource (operation handlers), data source and high level schema with validations

3. **Which issue(s) this PR fixes**

        (optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged):

        Fixes #OLYMP-36330

4. **Additional information**

5. **Special notes for your reviewer**

### CODE REVIEW

File: internal/resources/kubernetessecret/constants.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
+/*
+Copyright © 2023 VMware, Inc. All Rights Reserved.
+SPDX-License-Identifier: MPL-2.0
+*/
+
+package kubernetessecret
+
+const (
+	NameKey                       = "name"
+	NamespaceNameKey              = "namespace_name"
```
Comment (Vasundharashukla) [inline internal/resources/kubernetessecret/constants.go:10]: All these are globally exposed, do we need them outside this package?

File: internal/resources/kubernetessecret/data_source_secret.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	}
+
+	if d.Get(ExportKey).(bool) {
+		secretexportResp, err = config.TMCConnection.SecretExportResourceService.SecretExportResourceServiceGet(constructFullnameSecetExport(d))
+		if err != nil || secretexportResp == nil {
+			return diag.FromErr(errors.Wrapf(err, "unable to get Tanzu Mission Control secret export entry, name : %s", secretName))
+		}
+	}
+
+	status["export_phase"] = secretexportResp.SecretExport.Status.Conditions["Ready"]
```
Comment (Vasundharashukla) [inline internal/resources/kubernetessecret/data_source_secret.go:61]: Should we have `export_phase` assigned to a key in constants?

File: internal/resources/kubernetessecret/resource_secret.go [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+	"github.com/vmware/terraform-provider-tanzu-mission-control/internal/helper"
+	secretmodel "github.com/vmware/terraform-provider-tanzu-mission-control/internal/models/kubernetessecret/cluster"
+	secretexportmodel "github.com/vmware/terraform-provider-tanzu-mission-control/internal/models/kubernetessecret/cluster/secretexport"
+	"github.com/vmware/terraform-provider-tanzu-mission-control/internal/resources/common"
+	"github.com/vmware/terraform-provider-tanzu-mission-control/internal/resources/kubernetessecret/scope"
+)
+
+func ResourceSecret() *schema.Resource {
+	return &schema.Resource{
+		ReadContext:   dataSourceSecretRead,
```
Comment (Vasundharashukla) [inline internal/resources/kubernetessecret/resource_secret.go:30]: Update  is not supported for Secret?

File: internal/resources/kubernetessecret/resource_secret.go [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+		CreateContext: resourceSecretCreate,
+		DeleteContext: resourceSecretDelete,
+		Schema:        secretSchema,
+	}
+}
+
+var secretSchema = map[string]*schema.Schema{
+	NameKey: {
```
Comment (Vasundharashukla) [inline internal/resources/kubernetessecret/resource_secret.go:38]: Add descriptions to all the schema components.

File: internal/resources/kubernetessecret/resource_secret.go [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+		Type:     schema.TypeString,
+		Required: true,
+		ForceNew: true,
+	},
+	NamespaceNameKey: {
+		Type:     schema.TypeString,
+		Required: true,
+		ForceNew: true,
+	},
+	OrgIdKey: {
```
Comment (Vasundharashukla) [inline internal/resources/kubernetessecret/resource_secret.go:48]: Add optional/required/computed etc. specifications

File: internal/resources/kubernetessecret/resource_secret.go [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+	MaxItems:    1,
+	Elem: &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			DockerConfigjsonKey: {
+				Type:        schema.TypeList,
+				Description: "SecretType definition - SECRET_TYPE_DOCKERCONFIGJSON, Kubernetes secrets type.",
+				MaxItems:    1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						UsernameKey: {
```
Comment (Vasundharashukla) [inline internal/resources/kubernetessecret/resource_secret.go:81]: Similarly, define optional/required/computed etc and other schema components for each key.

**Review by Vasundharashukla — CHANGES_REQUESTED** [review]

File: internal/resources/kubernetessecret/resource_secret.go (hunk 3 above)
Comment (Shubbhang351) [reply internal/resources/kubernetessecret/resource_secret.go:30]: updated

File: internal/resources/kubernetessecret/resource_secret.go (hunk 6 above)
Comment (Shubbhang351) [reply internal/resources/kubernetessecret/resource_secret.go:86]: added

File: internal/resources/kubernetessecret/resource_secret.go (hunk 5 above)
Comment (Shubbhang351) [reply internal/resources/kubernetessecret/resource_secret.go:51]: added

File: internal/resources/kubernetessecret/resource_secret.go (hunk 4 above)
Comment (Shubbhang351) [reply internal/resources/kubernetessecret/resource_secret.go:39]: added

File: internal/resources/kubernetessecret/data_source_secret.go (hunk 2 above)
Comment (Shubbhang351) [reply internal/resources/kubernetessecret/data_source_secret.go]: updated

File: internal/resources/kubernetessecret/resource_secret.go (hunk 5 above)
Comment (Vasundharashukla) [reply internal/resources/kubernetessecret/resource_secret.go:51]: have you added for `OrgIDKey`?

File: internal/resources/kubernetessecret/resource_secret.go [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+							Type:        schema.TypeString,
+							Required:    true,
+							Description: "Username of the registry.",
+						},
+						PasswordKey: {
```
Comment (Vasundharashukla) [inline internal/resources/kubernetessecret/resource_secret.go:91]: This should be marked as `Sensitive`

**Review by Vasundharashukla — CHANGES_REQUESTED** [review]

File: internal/resources/kubernetessecret/resource_secret.go (hunk 7 above)
Comment (Shubbhang351) [reply internal/resources/kubernetessecret/resource_secret.go:92]: updated PTAL

File: internal/resources/kubernetessecret/data_source_secret.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+package kubernetessecret
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/pkg/errors"
+	"github.com/vmware/terraform-provider-tanzu-mission-control/internal/authctx"
```
Comment (shreyassreenivas) [inline internal/resources/kubernetessecret/data_source_secret.go:14]: Three block import style

File: internal/resources/kubernetessecret/data_source_secret.go [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+	}
+
+	if d.Get(ExportKey).(bool) {
+		secretexportResp, err = config.TMCConnection.SecretExportResourceService.SecretExportResourceServiceGet(constructFullnameSecetExport(d))
+		if err != nil || secretexportResp == nil {
+			return diag.FromErr(errors.Wrapf(err, "unable to get Tanzu Mission Control secret export entry, name : %s", secretName))
+		}
+	}
+
+	status[SecretExportPhaseKey] = secretexportResp.SecretExport.Status.Conditions["Ready"]
```
Comment (shreyassreenivas) [inline internal/resources/kubernetessecret/data_source_secret.go:61]: this should be called inside the if condition right

File: internal/resources/kubernetessecret/resource_secret.go [hunk 10]
```diff
@@ -199,3 +221,187 @@ func getDecodedSpecData(data strfmt.Base64) (*dockerConfigJSON, error) {
...
+
+	secretResponse, err := config.TMCConnection.SecretResourceService.SecretResourceServiceCreate(secretRequest)
+
+	if err != nil {
+		return diag.FromErr(errors.Wrapf(err, "unable to create Tanzu Mission Control secret entry, name : %s", NameKey))
+	}
+
+	d.SetId(secretResponse.Secret.Meta.UID)
+
+	export := d.Get(ExportKey).(bool)
```
Comment (shreyassreenivas) [inline internal/resources/kubernetessecret/resource_secret.go:282]: this can be used in the if condition itself

File: internal/resources/kubernetessecret/resource_secret.go [hunk 11]
```diff
@@ -199,3 +221,187 @@ func getDecodedSpecData(data strfmt.Base64) (*dockerConfigJSON, error) {
...
+
+func resourceSecretInPlaceUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) (diags diag.Diagnostics) {
+	config := m.(authctx.TanzuContext)
+
+	updateRequired := false
+
+	switch {
+	case common.HasMetaChanged(d):
+		fallthrough
+	case d.HasChange(helper.GetFirstElementOf(specKey, DockerConfigjsonKey, UsernameKey, PasswordKey, ImageRegistryURLKey)) || d.HasChange(ExportKey):
```
Comment (shreyassreenivas) [inline internal/resources/kubernetessecret/resource_secret.go:335]: check this condition

File: internal/resources/kubernetessecret/resource_secret.go [hunk 12]
```diff
@@ -199,3 +221,187 @@ func getDecodedSpecData(data strfmt.Base64) (*dockerConfigJSON, error) {
...
+			secretexportRequest := &secretexportmodel.VmwareTanzuManageV1alpha1ClusterNamespaceSecretExportRequest{
+				SecretExport: &secretexportmodel.VmwareTanzuManageV1alpha1ClusterNamespaceSecretExport{
+					FullName: constructFullnameSecetExport(d),
+					Meta:     common.ConstructMeta(d),
+				},
+			}
+
+			_, err := config.TMCConnection.SecretExportResourceService.SecretExportResourceServiceCreate(secretexportRequest)
+
+			if err != nil {
```
Comment (shreyassreenivas) [inline internal/resources/kubernetessecret/resource_secret.go:395]: you should skip already exists error

**Comment by Vasundharashukla** [comment]
@Shubbhang351 kindly resolve the conflicts and also rebase your branch, this PR has some of the merged commit changes.

**Comment by Shubbhang351** [comment]
add resource, 
Screenshot 2023-03-09 at 9 49 37 PM

**Comment by Shubbhang351** [comment]
No diff
Screenshot 2023-03-09 at 10 20 40 PM

**Comment by Shubbhang351** [comment]
delete resource
Screenshot 2023-03-09 at 10 22 50 PM

File: internal/resources/kubernetessecret/resource_secret.go [hunk 13]
```diff
@@ -35,13 +47,16 @@ var secretSchema = map[string]*schema.Schema{
 		ForceNew: true,
 	},
 	OrgIDKey: {
-		Type: schema.TypeString,
+		Type:     schema.TypeString,
```
Comment (Vasundharashukla) [inline internal/resources/kubernetessecret/resource_secret.go:50]: Description

File: internal/resources/kubernetessecret/resource_secret.go [hunk 14]
```diff
@@ -35,13 +47,16 @@ var secretSchema = map[string]*schema.Schema{
...
+		Optional: true,
 	},
 	scope.ScopeKey: scope.ScopeSchema,
 	specKey:        secretSpec,
+	common.MetaKey: common.Meta,
 	ExportKey: {
 		Type:     schema.TypeBool,
-		Required: true,
+		Optional: true,
+		Default:  false,
```
Comment (Vasundharashukla) [inline internal/resources/kubernetessecret/resource_secret.go:59]: Description

**Review by Vasundharashukla — CHANGES_REQUESTED** [review]

File: internal/resources/kubernetessecret/resource_secret.go (hunk 13 above)
Comment (Shubbhang351) [reply internal/resources/kubernetessecret/resource_secret.go]: Updated PTAL

File: internal/resources/kubernetessecret/resource_secret.go (hunk 12 above)
Comment (Shubbhang351) [reply internal/resources/kubernetessecret/resource_secret.go:383]: Updated PTAL

File: internal/resources/kubernetessecret/resource_secret.go (hunk 11 above)
Comment (Shubbhang351) [reply internal/resources/kubernetessecret/resource_secret.go]: Updated PTAL

File: internal/resources/kubernetessecret/resource_secret.go (hunk 10 above)
Comment (Shubbhang351) [reply internal/resources/kubernetessecret/resource_secret.go]: Updated PTAL

File: internal/resources/kubernetessecret/data_source_secret.go (hunk 9 above)
Comment (Shubbhang351) [reply internal/resources/kubernetessecret/data_source_secret.go]: Updated PTAL

File: internal/resources/kubernetessecret/data_source_secret.go (hunk 8 above)
Comment (Shubbhang351) [reply internal/resources/kubernetessecret/data_source_secret.go:15]: Updated PTAL

File: internal/resources/kubernetessecret/resource_secret.go [hunk 15]
```diff
@@ -6,51 +6,78 @@ SPDX-License-Identifier: MPL-2.0
...
+	"github.com/vmware/terraform-provider-tanzu-mission-control/internal/helper"
 	secretmodel "github.com/vmware/terraform-provider-tanzu-mission-control/internal/models/kubernetessecret/cluster"
+	secretexportmodel "github.com/vmware/terraform-provider-tanzu-mission-control/internal/models/kubernetessecret/cluster/secretexport"
+	"github.com/vmware/terraform-provider-tanzu-mission-control/internal/resources/common"
 	"github.com/vmware/terraform-provider-tanzu-mission-control/internal/resources/kubernetessecret/scope"
 )
 
+type (
+	contextMethodKey struct{}
+)
```
Comment (ishangupta-ds) [inline internal/resources/kubernetessecret/resource_secret.go:31]:
```suggestion
type contextMethodKey struct{}
```

File: internal/resources/kubernetessecret/resource_secret.go [hunk 16]
```diff
@@ -6,51 +6,78 @@ SPDX-License-Identifier: MPL-2.0
...
 	return &schema.Resource{
-		Schema: secretSchema,
+		CreateContext: resourceSecretCreate,
+		DeleteContext: resourceSecretDelete,
+		UpdateContext: resourceSecretInPlaceUpdate,
+		ReadContext:   dataSourceSecretRead,
+		Schema:        secretSchema,
+		CustomizeDiff: customdiff.All(
+			validateSpec,
+		),
```
Comment (ishangupta-ds) [inline internal/resources/kubernetessecret/resource_secret.go:42]:
```suggestion
		CustomizeDiff: validateSpec,
```

File: internal/resources/kubernetessecret/resource_secret.go [hunk 17]
```diff
@@ -6,51 +6,78 @@ SPDX-License-Identifier: MPL-2.0
...
 
 var secretSchema = map[string]*schema.Schema{
 	NameKey: {
-		Type:     schema.TypeString,
-		Required: true,
-		ForceNew: true,
+		Type:        schema.TypeString,
+		Description: "Name of the secret resource.",
+		Required:    true,
+		ForceNew:    true,
```
Comment (ishangupta-ds) [inline internal/resources/kubernetessecret/resource_secret.go:51]: nit: Please add any validations/constraints that have been specified in the swagger file. Ex - length of the name, the format etc.

File: internal/resources/kubernetessecret/resource_secret.go [hunk 18]
```diff
@@ -60,25 +87,30 @@ var secretSpec = &schema.Schema{
 	Description: "Spec for the kubernetes secret",
 	Optional:    true,
 	MaxItems:    1,
+	MinItems:    1,
 	Elem: &schema.Resource{
 		Schema: map[string]*schema.Schema{
```
Comment (ishangupta-ds) [inline internal/resources/kubernetessecret/resource_secret.go:92]: nit: Please add any validations/constraints that have been specified in the swagger file. This applies to the entire schema.

File: internal/resources/kubernetessecret/resource_secret.go [hunk 19]
```diff
@@ -87,7 +119,19 @@ var secretSpec = &schema.Schema{
...
+	case (diff.Get(helper.GetFirstElementOf(specKey, DockerConfigjsonKey, ImageRegistryURLKey))).(string) == "":
+		return fmt.Errorf("invalid value for %v, can not be nil string", ImageRegistryURLKey)
+	case (diff.Get(helper.GetFirstElementOf(specKey, DockerConfigjsonKey, UsernameKey))).(string) == "":
+		return fmt.Errorf("invalid value for %v, can not be nil string", UsernameKey)
+	case (diff.Get(helper.GetFirstElementOf(specKey, DockerConfigjsonKey, PasswordKey))).(string) == "":
+		return fmt.Errorf("invalid value for %v, can not be nil string", PasswordKey)
+	default:
+		return nil
+	}
+}
```
Comment (ishangupta-ds) [inline internal/resources/kubernetessecret/resource_secret.go:133]: Why is this needed? Can't we add schema validations instead?

Explore adding validation to the different blocks and attributes in the schema.

File: internal/resources/kubernetessecret/resource_secret.go [hunk 20]
```diff
@@ -208,3 +251,169 @@ func getDecodedSpecData(data strfmt.Base64) (*dockerConfigJSON, error) {
...
+	fullname = (scope.ConstructScope(d, name)).FullnameCluster
+
+	fullname.NamespaceName, _ = d.Get(NamespaceNameKey).(string)
+
+	fullname.OrgID, _ = d.Get(OrgIDKey).(string)
+
+	return fullname
+}
+
+func constructFullnameSecetExport(d *schema.ResourceData) (fullname *secretexportmodel.VmwareTanzuManageV1alpha1ClusterNamespaceSecretexportFullName) {
```
Comment (ishangupta-ds) [inline internal/resources/kubernetessecret/resource_secret.go:267]: Please create a file called - `fullname.go`

**Review by ishangupta-ds — CHANGES_REQUESTED** [review]

File: internal/resources/kubernetessecret/resource_secret.go (hunk 19 above)
Comment (Shubbhang351) [reply internal/resources/kubernetessecret/resource_secret.go]: removed

File: internal/resources/kubernetessecret/resource_secret.go (hunk 18 above)
Comment (Shubbhang351) [reply internal/resources/kubernetessecret/resource_secret.go:96]: added PTAL

File: internal/resources/kubernetessecret/resource_secret.go (hunk 17 above)
Comment (Shubbhang351) [reply internal/resources/kubernetessecret/resource_secret.go:46]: added PTAL

File: internal/resources/kubernetessecret/resource_secret.go (hunk 16 above)
Comment (Shubbhang351) [reply internal/resources/kubernetessecret/resource_secret.go]: updated PTAL

File: internal/resources/kubernetessecret/resource_secret.go (hunk 15 above)
Comment (Shubbhang351) [reply internal/resources/kubernetessecret/resource_secret.go]: Updated PTAL

File: internal/resources/kubernetessecret/data_source_secret.go [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+		secretexportResp *secretexportclustermodel.VmwareTanzuManageV1alpha1ClusterNamespaceGetSecretExportResponse
+		err              error
+	)
+
+	secretName, _ := d.Get(NameKey).(string)
+
+	secretResp, err = config.TMCConnection.SecretResourceService.SecretResourceServiceGet(constructFullname(d))
+	if err != nil || secretResp == nil {
+		if clienterrors.IsNotFoundError(err) {
+			if ctx.Value(contextMethodKey{}) == "dataSourceRead" {
```
Comment (shreyassreenivas) [inline internal/resources/kubernetessecret/data_source_secret.go:47]: make dataSourceRead into a constant

File: internal/resources/kubernetessecret/data_source_secret.go [hunk 22]
```diff
@@ -0,0 +1,108 @@
...
+
+			_ = schema.RemoveFromState(d, m)
+
+			return diags
+		}
+
+		return diag.FromErr(errors.Wrapf(err, "unable to get Tanzu Mission Control secret entry, name : %s", secretName))
+	}
+
+	var pass string
```
Comment (shreyassreenivas) [inline internal/resources/kubernetessecret/data_source_secret.go:59]: rename it to passwd

File: internal/resources/kubernetessecret/data_source_secret.go [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+					return diag.FromErr(err)
+				}
+			default:
+				return diag.FromErr(errors.Wrapf(err, "unable to get Tanzu Mission Control secret export entry, name : %s", secretName))
+			}
+		} else {
+			status[SecretExportPhaseKey] = secretexportResp.SecretExport.Status.Conditions["Ready"].Reason
+		}
+	} else {
+		secretexportResp, err = config.TMCConnection.SecretExportResourceService.SecretExportResourceServiceGet(constructFullnameSecetExport(d))
```
Comment (shreyassreenivas) [inline internal/resources/kubernetessecret/data_source_secret.go:84]: why is the call being done in if and else ?

File: internal/resources/kubernetessecret/data_source_secret.go (hunk 21 above)
Comment (Shubbhang351) [reply internal/resources/kubernetessecret/data_source_secret.go]: update PTAL

File: internal/resources/kubernetessecret/data_source_secret.go [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+	var (
+		diags            diag.Diagnostics
+		secretResp       *secretclustermodel.VmwareTanzuManageV1alpha1ClusterNamespaceGetSecretResponse
+		secretexportResp *secretexportclustermodel.VmwareTanzuManageV1alpha1ClusterNamespaceGetSecretExportResponse
+		err              error
+	)
+
+	secretName, _ := d.Get(NameKey).(string)
+
+	pass, _ := (d.Get(helper.GetFirstElementOf(specKey, DockerConfigjsonKey, PasswordKey))).(string)
```
[the remark this answers is not in the document]
Comment (Shubbhang351) [reply internal/resources/kubernetessecret/data_source_secret.go]: update PTAL

**Review by ishangupta-ds — APPROVED** [review]
LGTM!

File: internal/resources/kubernetessecret/data_source_secret.go (hunk 23 above)
Comment (Shubbhang351) [reply internal/resources/kubernetessecret/data_source_secret.go]: update PTAL

**Comment by Shubbhang351** [comment]
@shreyassreenivas please take a look

File: internal/resources/kubernetessecret/resource_secret.go [hunk 25]
```diff
@@ -208,3 +257,137 @@ func getDecodedSpecData(data strfmt.Base64) (*dockerConfigJSON, error) {
...
+			meta.Labels[common.CreatorLabelKey] = value
+		}
+
+		getResp.Secret.Meta.Labels = meta.Labels
+		getResp.Secret.Meta.Description = meta.Description
+	}
+
+	getResp.Secret.Spec = constructSpec(d)
+
+	_, errs := config.TMCConnection.SecretResourceService.SecretResourceServiceUpdate(
```
Comment (shreyassreenivas) [inline internal/resources/kubernetessecret/resource_secret.go:359]: why errs ?

**Review by shreyassreenivas — APPROVED** [review]

File: internal/resources/kubernetessecret/resource_secret.go (hunk 25 above)
Comment (shreyassreenivas) [reply internal/resources/kubernetessecret/resource_secret.go:359]: address this

Comment (Shubbhang351) [reply internal/resources/kubernetessecret/resource_secret.go]: updated PTAL

**Review by Vasundharashukla — APPROVED** [review]
